### PR TITLE
Improve the foundation of the visualizer

### PR DESCRIPTION
**Pull request by qgolsteyn**
### Description
This PR merges the work I did on the visualizer over the winter break. This includes the addition of a Redux library to manage the application state, as well as a completely new approach to the UI.

This PR sets the foundation for future PRs. Most of the code will still be revisited but I think I managed to get the architecture specified.

### Testing Done
Manual testing

### Length Justification
This has grown organically throughout the break and was difficult to break down as I was trying to reach feature parity with the previous version of the visualizer. I believe the code is now more stable and future PR will be kept small.

### Review Checklist
*(Please check every item to indicate your code complies with it (by changing `[ ]`->`[x]`). This will hopefully save both you and the reviewer(s) a lot of time!)*
***It is the reviewers responsibility to also make sure every item here has been covered***
- [X] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [X] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom` 
- [X] **Remove all commented out code**
- [X] **Remove extra print statements**: for example, those just used for testing
- [X] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [X] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

***Feel free to make additions of things that we should be checking to this file if you think there's something missing!!!!***

### CODE REVIEW

**Review by garethellis0 — COMMENTED** [review]
The below is just a dump of some thoughts after a conversation with Quentin, will actually review tmw/monday:

- can the logo just be raw html instead of a portal?
- should add explanation for: when "html" is really not html (it's JSX!), where the actual (_high level_) layout is defined (it's in CSS!!), that each "puzzle piece" in the CSS grid is really just a React component, 
- define that the HTML really just specifies what is on the screen (just some id's really)(**there is _really_ no work done in the HTML), then CSS defines what part of the screen each ID takes up and roughly what it looks like. Then **inside** each of these boxes (which have an ID) there is a react componenet that actually defines what appears in them
- should talk about design rationale (from top down) behind portals: to hold to a global layout while still being able to swap components within that layout without having to worry about maintaining the layout outside of a single file

File: src/corner_kick/src/store/actions/ros.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
+/**
+ * This file specifies ROS specific action
+ */
+
+import { createAction } from 'typesafe-actions';
+
+/**
+ * Connect to ROS
+ */
+export const start = createAction('ros/START');
```
Comment (garethellis0) [inline src/corner_kick/src/store/actions/ros.ts]: The `ros/` scoping is implicit, not explicit, so I _think_ `ros_START` would be more clear (also in _most_ languages, `/` is not allowed in id's)

**Review by garethellis0 — COMMENTED** [review]
Quick explanation of some functional bits would be good.

File: src/corner_kick/src/containers/mainUI/GlobalStyle.tsx [hunk 2]
```diff
@@ -0,0 +1,186 @@
+/**
+ * This file defines the main styling of the application and connects the
+ * theme to all styled-components instances
```
Comment (garethellis0) [inline src/corner_kick/src/containers/mainUI/GlobalStyle.tsx]: Quick note that this is where the overall panel "layout" is defined would be good here.

File: src/corner_kick/src/containers/mainUI/GlobalStyle.tsx [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+                info: settings.colors_info,
+                debug: settings.colors_debug,
+
+                red: settings.colors_red,
+                orange: settings.colors_orange,
+                yellow: settings.colors_yellow,
+                green: settings.colors_green,
+                cyan: settings.colors_cyan,
+                blue: settings.colors_blue,
+                purple: settings.colors_purple,
```
Comment (garethellis0) [inline src/corner_kick/src/containers/mainUI/GlobalStyle.tsx]: This may be a question for when we do layering, but if I wanted to add a new color, how would I know everywhere to look? These seem somewhat scattered right now.

File: src/corner_kick/src/types/theme.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    panel: string;
+    border: string;
+
+    success: string;
+    error: string;
+    warn: string;
+    info: string;
+    debug: string;
+
+    red: string;
```
Comment (garethellis0) [inline src/corner_kick/src/types/theme.ts]: Bit confused about this; seems like `red` should be a value for `fg`, `bg`, etc.? Why are they defined at the same level here?

**Review by garethellis0 — COMMENTED** [review]
Overall this seems Ok, but I'm not sure I fully understand everything just yet. Main comment would be to make sure that all functions have an appropriate javadoc style comment to help readability (especially given that this will be read a lot by people who don't really know web dev)

File: src/corner_kick/src/store/actions/ros.ts (hunk 1 above)
Comment (qgolsteyn) [reply src/corner_kick/src/store/actions/ros.ts]: Technically specific to Redux, but I implemented the change

File: src/corner_kick/src/containers/mainUI/GlobalStyle.tsx (hunk 3 above)
Comment (qgolsteyn) [reply src/corner_kick/src/containers/mainUI/GlobalStyle.tsx]: That's fair, let's revisit that when we do layering. I added that requirement to #184

File: src/corner_kick/src/containers/mainUI/GlobalStyle.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (qgolsteyn) [reply src/corner_kick/src/containers/mainUI/GlobalStyle.tsx]: Separated layout from theming. Layout is now defined in [style.css](https://github.com/qgolsteyn/Software/blob/master/src/corner_kick/src/style.css) and theming is defined in [Theme.tsx](https://github.com/qgolsteyn/Software/blob/master/src/corner_kick/src/containers/mainUI/Theme.tsx).

**Comment by qgolsteyn** [comment]
@garethellis0 I think I covered all the points

File: src/corner_kick/src/containers/mainUI/GlobalStyle.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/corner_kick/src/containers/mainUI/GlobalStyle.tsx]: I don't see `style.css` in this PR? Looks like that link goes to `master` on your fork, but this PR is from `destruction`?

Similairly for `Theme.tsx`

**Review by garethellis0 — COMMENTED** [review]
@qgolsteyn - are you sure you pushed your changes to this branch? Looks like the changes might've gone to your `master` branch instead of `destruction`

File: src/corner_kick/src/containers/mainUI/GlobalStyle.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/corner_kick/src/containers/mainUI/GlobalStyle.tsx]: Thoughts on changing `style.css` to `grid_layout.css` (or similar)? I think it would make more sense if one were trying to hunt down where the layout was.

File: src/corner_kick/src/types/theme.ts (hunk 4 above)
Comment (garethellis0) [reply src/corner_kick/src/types/theme.ts]: Hmm, some comments here would probably go a long way. If I understand this correctly though, could we change `red`, `orange`, etc. to `red_shade`, `orange_shade`, etc.?

Also, why are they strings? Seems like the shade of red should be defined as a number, or maybe an enum? 

Also, what are `fg` and `bg` then? Not rgb values if `red`, `orange`, etc define the shade?

And what if I set a shade of red and a shade of green?

Sorry for the comment dump above, I think that a few quick comments in the code here would likely address my questions, but just leaving the above brain dump to hopefully help convey where my confusion is coming from.

File: src/corner_kick/common.d.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+/**
+ * This file defines supported file extensions
+ */
+
+declare module '*.css';
+declare module '*.png';
+declare module '*.jpg';
+declare module '*.svg';
```
Comment (garethellis0) [inline src/corner_kick/common.d.ts:10]: What about `.ts`, `.tsx`, `.js`, etc.? Also why is this file needed in the first place?

File: src/corner_kick/config/webpack.base.config.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+                        loader: 'file-loader',
+                        options: {},
+                    },
+                ],
+            },
+        ],
+    },
+    resolve: {
+        // Make sure if you add a new file support to add its extension here.
+        extensions: ['.js', '.ts', '.tsx', '.json'],
```
Comment (garethellis0) [inline src/corner_kick/config/webpack.base.config.js:54]: what about `.css`, `.svg`, etc?

File: src/corner_kick/config/webpack.base.config.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    // Our project entry point.
+    entry: path.resolve(__dirname, '../src/index.ts'),
+
+    // We generate a bundle in the build folder
+    output: {
+        path: path.resolve(__dirname, '../build'),
+        filename: '[name].[contenthash].js',
+    },
+
+    // This target allows us to access DOM libraries simultaneously
```
Comment (garethellis0) [inline src/corner_kick/config/webpack.base.config.js]: Why do we wish to do this? (I can guess at performance, but generally comments should explicitly state _why_, rather then _what_ or _how_)

File: src/corner_kick/config/webpack.prod.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+                        // npm package names are URL-safe, but some servers don't like @ symbols
+                        return `npm.${packageName.replace('@', '')}`;
+                    },
+                },
+            },
+        },
+    },
+    // We run the visualizer locally so we can keep our bundles a bit bigger
+    // than the default max size
```
Comment (garethellis0) [inline src/corner_kick/config/webpack.prod.js]: Why is this a good thing to do?

File: src/corner_kick/src/App.tsx [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+import { Main } from './components/portals/Main';
+import { Sidebar } from './components/portals/Sidebar';
+import { SidebarTitle } from './components/portals/SidebarTitle';
+import { Theme } from './containers/mainUI/Theme';
+import { createStore } from './store';
+
+const store = createStore();
+
+// Top-level React component in the application
+export const App = () => (
```
Comment (garethellis0) [inline src/corner_kick/src/App.tsx:21]: I know this should be known and understood by anyone with a web dev background, but could we add a comment here saying something to the effect of `// This is *NOT* HTML code, it is ...`? Just to make it more clear for the likely case where people with minimal (at best) web experience come bumbling through here.

File: src/corner_kick/README.md [hunk 10]
```diff
@@ -1,6 +1,22 @@
-# Corner Kick
+<p align="center">
+  <img src="./src/assets/logo.svg" width="64px">
```
Comment (garethellis0) [inline src/corner_kick/README.md:2]: I think this image is broken?

File: src/corner_kick/README.md [hunk 11]
```diff
@@ -1,6 +1,22 @@
...
+  <br>
+  <h1 align="center">Corner Kick</h1>
+</p>
 
-Corner Kick is an Electron app that allows for the visualization of UBC Thunderbots' simulator.
+Corner Kick is a web app that allows for the visualization of UBC Thunderbots' AI.
+
+![application screenshot](screenshot.png)
+
+**Table of content**
```
Comment (garethellis0) [inline src/corner_kick/README.md:11]: Could we add a small section explaining portals? Especially given that these are something new, custom, and specific to our team.

File: src/corner_kick/src/index.html [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+            href="https://fonts.googleapis.com/icon?family=Material+Icons"
+            rel="stylesheet"
+        />
+    </head>
+
+    <body>
+        <!-- This is where our application is initially mounted -->
+        <div id="react"></div>
+
+        <!-- This where various UI elements are added using React portals-->
```
Comment (garethellis0) [inline src/corner_kick/src/index.html:36]: Could we add a quick note here to the effect of `The order/arrangement of these elements should have no effect *whatsoever* on the layout of the actual applications, these are just "hangers" to allow us to get a reference on which to hang various components`

As another thought, why do we need multiple "hangers"? Could we not just have one hanger here to hang everything on? A quick comment elaborating on this would be good.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Sorry for another round of comments, this is a large enough PR that I definitely feel I need to have a few passes with solid breaks in between to really get it.

- A few of my comments are about making sure comments in code indicate _why_ we are doing something (though I don't super object to also indicating _what_ we're doing or _how_ we're doing it)

File: src/corner_kick/src/containers/mainUI/GlobalStyle.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (qgolsteyn) [reply src/corner_kick/src/containers/mainUI/GlobalStyle.tsx]: Moved layout to `index.html`

File: src/corner_kick/README.md (hunk 10 above)
Comment (qgolsteyn) [reply src/corner_kick/README.md:2]: Fixed

File: src/corner_kick/config/webpack.base.config.js (hunk 6 above)
Comment (qgolsteyn) [reply src/corner_kick/config/webpack.base.config.js:54]: Clarified in a comment `These are the extensions that Webpack will try to resolve (aka transpile into a .js bundle)`

File: src/corner_kick/common.d.ts (hunk 5 above)
Comment (qgolsteyn) [reply src/corner_kick/common.d.ts:10]: Clarified into a comment
```
/**
 * This file defines additional extensions that could be present in Typescipt code and
 * are supported by our Webpack configuration. Without defining them, Typescript will
 * throw a transpile time error when we try to import a file of this type.
 */
```

File: src/corner_kick/src/App.tsx (hunk 9 above)
Comment (qgolsteyn) [reply src/corner_kick/src/App.tsx:21]: Done
```
// Top-level React component in the application
// !!!! THIS ... IS ... NOT ... HTML !!!!
// IF YOU DIDN'T REALIZE THIS, PLEASE GO AND READ
// https://reactjs.org/docs/introducing-jsx.html
```

File: src/corner_kick/src/index.html (hunk 12 above)
Comment (qgolsteyn) [reply src/corner_kick/src/index.html:36]: Added comment. We need multiple hangers because we cannot define a complex layout using only one `div` element

File: src/corner_kick/config/webpack.base.config.js (hunk 7 above)
Comment (qgolsteyn) [reply src/corner_kick/config/webpack.base.config.js]: Added clarification.
```
// We use the web target to give us access to DOM-related JS functions
// A good analogy is selecting the STD library used in a C/C++ project
target: 'web',
```

File: src/corner_kick/README.md (hunk 11 above)
Comment (qgolsteyn) [reply src/corner_kick/README.md:11]: #### React Portals

Access to those section in the React application is done via [React Portals](https://reactjs.org/docs/portals.html).
Portals are used to ensure the application layout is enforced throughout all the pages while offering an
easy way to access them.

In essence then, the sections of the application are boxes where the React application can insert React components into.
This allows us to insert and swap UI elements in the application without worrying about the global layout.

Comment (qgolsteyn) [reply src/corner_kick/README.md:11]: This is what I have for that

File: src/corner_kick/config/webpack.prod.js (hunk 8 above)
Comment (qgolsteyn) [reply src/corner_kick/config/webpack.prod.js]: Added more clarification. We could make the bundle smaller through various performance fix, but
would make the build file more complex. Because we run locally, we do not have to worry too much about bundle size.

```
// Webpack complains that our bundle is too large.
// We run the visualizer locally so we can keep our bundles bigger
// than the default max size.
```

File: src/corner_kick/src/types/theme.ts (hunk 4 above)
Comment (qgolsteyn) [reply src/corner_kick/src/types/theme.ts]:
```
// This is the application's color palette. Colors need
// to be strings as this is what our CSS code accepts.
// Some of those attributes define color for a purpose
// (`fg` defines the color for a foreground elements - like text)
// while others define shades of a particular color (red defines the
// shade of red we use)
```

**Comment by qgolsteyn** [comment]
@garethellis0 Addressed all changes

File: src/corner_kick/README.md (hunk 11 above)
Comment (garethellis0) [reply src/corner_kick/README.md:11]: OH! I totally misunderstood when you explained this to me. I thought this was a totally new concept that you had come up with. My bad. Either way, this section is definitely good, so let's leave it!

File: src/corner_kick/src/index.html (hunk 12 above)
Comment (garethellis0) [reply src/corner_kick/src/index.html:36]: This makes a lot more sense with the layout being specified in `index.html`.

File: src/corner_kick/src/assets/logo.svg [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+            32H28V32C30.2091 
+            32 
+            32 
+            30.2091 
+            32 
+            28V28V2Z
+        " 
+        fill="#c2d94c"/>
+    <circle cx="13" cy="13" r="9" fill="#c2d94c"/>
+</svg>
```
Comment (garethellis0) [inline src/corner_kick/src/assets/logo.svg]: nit; github complaining about end-of-file

File: src/corner_kick/README.md [hunk 14]
```diff
@@ -36,20 +53,60 @@ The project is defined as follow:
...
-            -   `ui` - contains repeatable UI components
-        -   `modules` - contains high-level UI components (logger, visualizer, etc.)
-    -   `shared` - contains code shared by the two process
-        -   `ros` - contains the service communicating with ROS
-            -   `types` - contains ROS message type schema
+    -   `assets` - contains image assets
+    -   `components` - contains repeatable UI components
+    -   `containers` - contains high-level UI components (logger, visualizer, etc.)
+    -   `constants` - contains fixed values used throughout the application
+    -   `store` - contains logic for the application Redux state
```
Comment (garethellis0) [inline src/corner_kick/README.md:61]: Could we add a sub-entry here for `reducers` with a quick description and this link: https://redux.js.org/basics/reducers?

File: src/corner_kick/src/types/theme.ts [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * The application theme
+ */
+export interface IThemeProvider {
+    colors: IColorProvider;
+}
+
+/**
+ * The color available in the application
```
Comment (garethellis0) [inline src/corner_kick/src/types/theme.ts]:
``` suggestion
 * The colors available in the application
```

**Review by garethellis0 — DISMISSED** [review]
Left a few comments, but all small stuff. Once those are resolved this LGTM :+1: :tada:. Great, great work getting this pushed through (and tolerating my bumbling non-webdev reviews :joy: ).

File: src/corner_kick/src/assets/logo.svg (hunk 13 above)
Comment (qgolsteyn) [reply src/corner_kick/src/assets/logo.svg]:
``` suggestion
</svg>

```

File: src/corner_kick/README.md (hunk 14 above)
Comment (qgolsteyn) [reply src/corner_kick/README.md:61]: #### Reducer functions and Redux saga

Reducer functions are the cornerstone of the Redux approach. Once a Redux action is dispatched, reducer functions
determine what the new state will look like as a result of the action sent. Reducer functions do not permit for
any side-effects to ensure a deterministic behaviour. For a state and action, the same new state should be returned
by the reducer funtion. This, however, does not permit for more complex logic, such as communicating with ROS,
to occur. To allow for this, we introduce sagas, which can intercept and act on Redux action. They can conduct some
business logic work, and reflect the result of this work in the state by emmiting a Redux action when needed.

For more information about reducer functions, read the [official Redux doc on that topic](https://redux.js.org/basics/reducers).
For more information about Redux sagas, read the [official Redux saga doc](https://redux-saga.js.org/).

**Comment by qgolsteyn** [comment]
@MathewMacDougall your turn to review

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
This took a while to get through, but I don't have any major comments to add. The comments that exist combined with my super-basic knowledge of JS and React are enough for me to get a rough idea of how things work and are connected.

Just 2 small questions but otherwise this looks good.

File: src/corner_kick/src/store/sagas/ros.ts [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+
+let ros: ROSLIB.Ros | null = null;
+const rosChannel = channel();
+
+/**
+ * Function first called when the application first starts
+ */
+export default function* init() {
+    // Listen to start actions and start ROS
+    yield takeLatest('ros/START', startROS);
```
[the remark this answers is not in the document]
Comment (qgolsteyn) [reply src/corner_kick/src/store/sagas/ros.ts]: GREAT catch. That would be a bug. Hopefully our next PR, introducing a unit testing framework, will avoid these kinds of bugs. I modified the code to avoid these kinds of bugs in the future.

File: src/corner_kick/src/store/index.ts [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Create a Redux store and attaches Chrome debugging tools
+ */
+export const createStore = () => {
+    // We use Redux Saga for our application logic. It acts as a middleware in order to
+    // interact with our Redux application
+    const reduxSaga = createSagaMiddleware();
+
+    // We use the Redux debug Chrome extension to see what happens inside our Redux store
```
[the remark this answers is not in the document]
Comment (qgolsteyn) [reply src/corner_kick/src/store/index.ts]: Right, Chrome and Firefox actually. Modified the comment to specify that.

[the remark this answers is not in the document]
Comment (qgolsteyn) [reply src/corner_kick/src/store/index.ts]: Also added some info about this tool in the README

File: src/corner_kick/src/store/sagas/ros.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (qgolsteyn) [reply src/corner_kick/src/store/sagas/ros.ts]: Now getting the type directly from the action. Changing the type on the action will update sagas accordingly

**Comment by qgolsteyn** [comment]
@MathewMacDougall @garethellis0 Implemented all changes. Let me know if there is anything else that needs some modification.

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:
